hydra-node/src/Hydra/Logging.hs

```diff
@@ -95,15 +96,13 @@ withTracerOutputTo ::
   (Tracer m msg -> IO a) ->
   IO a
 withTracerOutputTo hdl namespace action = do
-  action tracer `finally` flushLogs
- where
-  tracer =
-    Tracer $
-      mkEnvelope namespace >=> liftIO . write . Aeson.encode
-
-  flushLogs = liftIO $ hFlush hdl
-
-  write bs = LBS.hPut hdl (bs <> "\n")
+  hdlMutex <- newMVar ()
+  let write bs = withMVar hdlMutex (\_ -> LBS.hPut hdl (bs <> "\n"))
+      flushLogs = liftIO $ hFlush hdl
+      tracer =
+        Tracer $
+          mkEnvelope namespace >=> liftIO . write . Aeson.encode
+   in action tracer `finally` flushLogs
```

Comment:
I don't see why this would be any different to before. I would expect `hPut` to write the whole bytestring to the handle atomically (because there is also a non-blocking variant).

Reply:
I wonder if it's not to do with laziness 🤔 
`write` does 2 things: It appends a newline to the encoded bytestring and `hPut` it. But the bytestring is lazy so I can imagine that the  `<>` does not actually append and only construct chunks, which are then put one after another. This could explain the fact you see an interleaving of the messages but they are not mangled. 
Perhaps using strict BS would help? 
```
write bs = BS.hPut hdl (toStrict $ bs <> "\n")
```

Reply:
That sounds reasonable. I would suggest using `Aeson.encodeStrict` though.

Reply:
That would be great if it existed 😂

Reply:
@ch1bo 
>  I would expect hPut to write the whole bytestring to the handle atomically (because there is also a non-blocking variant).

Actually, hPut does documentation does not mention anything about atomicity. The non-blocking versus blocking versions of hPut are referring to more low-level _system-ish_ concerns.

The idea with the blocking version is to block the current thread until all the bytes have been written to the underlying buffers. That is, if, for any reason, the underlying buffers are full at some point while writing, the blocking version will wait for the buffer data to be flushed to some other lower level buffer/storage and then continue writing until all the data passed as a parameter have been written.

The non-blocking version will more write whatever it can to the underlying buffers and, as soon as something is full, instead of blocking, it will just return with the un-written data send back to the caller thread. In which case, the caller thread has to manage all the logic of re-submitting data later itself.

This is a big approximation, of course, but I think it's quite fair to describe the difference.

In neither of these two approaches is any atomicity or mutex involved. All the threads which call any of these functions on the same underlying file descriptor are in concurrence to access and write data to this file descriptor.

I think is it this concurrency which explains the mangling you've seen in the log outputs. I agree that you didn't see that much mangling, only _missing newlines_ but the problem with this un-determinism is that there is no way to really predict what will happen. Depending on how the Haskell thread scheduler works and depending on ho our application logs and its concurrency level, we might see very subtle/limited impact but, nevertheless, any code which would not put some sort of mutex to protect concurrent access to a file descriptor is unpredictable.